llvm/lib/Analysis/Loads.cpp

```diff
@@ -318,11 +321,24 @@ bool llvm::isDereferenceableAndAlignedInLoop(LoadInst *LI, Loop *L,
       // TODO: generalize if a case found which warrants
       if (Offset->getAPInt().urem(Alignment.value()) != 0)
         return false;
+      if (StepIsNegative) {
+        // In the last iteration of the loop the address we access we will be
+        // lower than the first by (TC - 1) * Step. So we need to make sure
+        // that there is enough room in Offset to accomodate this.
+        APInt SubOffset = (TC - 1) * AbsStep;
+        if (Offset->getAPInt().ult(SubOffset))
+          return false;
+        // We can safely use the new base because the decrementing pointer is
+        // always guaranteed to be >= new base. The total access size needs to
+        // take into account the start offset and the loaded element size.
+        AccessSize = Offset->getAPInt() + EltSize;
```

Comment:
Does this need an overflow check and bail out in case we have a narrow IV where incrementing by EltSize wraps? Alternatively we could probably also perform the computation in  always in something like i64?

Reply:
Good point! I've added the overflow check.

Reply:
Thanks for the update! Do you think it would be possible to add a test case that covers the overflow case (e.g. with an i8 induction that wraps when adding the element size?

Reply:
I tried doing that with an i8, but the offset return by SCEV is always 64-bit. So in practice I'd need a test with an alloca just less than UINT64_MAX bytes in size and an offset just less than UINT64_MAX, in order to exercise the overflow case. I wasn't sure if this was fragile, i.e. are allocas of such a large size even legal given that I don't think GEPs even support an offset greater than SINT64_MAX (since offsets are signed)? Having said that, I'm open to ideas about how to write a sensible test for this!

Reply:
Would something like the below work? I might be missing something, but it looks like the overflow check should trigger there?


```

target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:1"

declare void @init(ptr nocapture nofree)


define void @test(ptr nocapture noundef writeonly %dest) {

entry:
  %local_dest = alloca [128 x i8], align 4
  %local_src = alloca [128 x i8], align 4
  %local_cmp = alloca [128 x i8], align 4
  call void @init(ptr %local_src)
  call void @init(ptr %local_cmp)
  br label %for.body

for.body:
  %iv = phi i8 [ 255, %entry ], [ %iv.next, %for.inc ]
  %arrayidx = getelementptr inbounds [1024 x i8], ptr %local_cmp, i8 0, i8 %iv
  %0 = load i8, ptr %arrayidx, align 1
  %cmp3.not = icmp eq i8 %0, 3
  br i1 %cmp3.not, label %for.inc, label %if.then

if.then:
  %arrayidx5 = getelementptr inbounds [1024 x i8], ptr %local_src, i8 0, i8 %iv
  %1 = load i8, ptr %arrayidx5, align 1
  %mul = shl nsw i8 %1, 2
  %arrayidx7 = getelementptr inbounds [1024 x i8], ptr %local_dest, i8 0, i8 %iv
  store i8 %mul, ptr %arrayidx7, align 1
  br label %for.inc

for.inc:
  %iv.next = add nsw i8 %iv, -1
  %cmp2.not = icmp eq i8 %iv, 200
  br i1 %cmp2.not, label %for.cond.cleanup, label %for.body

for.cond.cleanup:
  call void @llvm.memcpy.p0.p0.i8(ptr noundef nonnull align 4 dereferenceable(1024) %dest, ptr noundef nonnull align 4 dereferenceable(1024) %local_dest, i8 1024, i1 false)
  ret void
}
```

Reply:
> Would something like the below work? I might be missing something, but it looks like the overflow check should trigger there?

Yep, that's exactly the test I tried and the offset produced was a 64-bit APInt. I'll try again with your IR just in case you've written it differently to me.

Reply:
Right, so the offset is coming out as 64-bits, but it's probably not what you were expecting - Offset = 0xFFFFFFFFFFFFFFFF. Again, I think it's because the offset in a GEP is treated as being signed so (i8 255) is equivalent to (i8 -1) and is being sign-extended to (i64 -1). So yes technically when we try to (unsigned) add the element size (1 byte) to the offset it overflows, but I'm not sure we can rely upon that for testing overflow behaviour.

This actually makes me wonder whether there are already problems with the existing code, which seems to unconditionally treat the offset as unsigned. For example, should this code

```
       // TODO: generalize if a case found which warrants
       if (Offset->getAPInt().urem(Alignment.value()) != 0)
         return false;
```

be rewritten to

```
       // TODO: generalize if a case found which warrants
       if (Offset->getAPInt().srem(Alignment.value()) != 0)
         return false;
```

etc?

Reply:
Or perhaps bail out entirely if the offset is negative?

Reply:
I think it might makes sense for now to bail out for negative values - it doesn't cause any existing tests to fail. It would also make it almost impossible to write an overflow test because you'd need something like an offset of 0x7FFFFFFFFFFFFFFF and an element size of 0x7FFFFFFFFFFFFFFF

Reply:
I think it makes sense to restrict this to non-negative values, thanks! It would probably make sense to split off the patch with the extra bail-out separately. Also adding @preames who added some of that code originally I think

Reply:
So I originally thought about splitting off the change in a separate patch, but I couldn't see any way of testing it by itself because the overflow checks catch this by accident. If you want I can split it off into a separate PR - there just won't be any tests for it.

Reply:
Created https://github.com/llvm/llvm-project/pull/99490 for the negative offset check.